python/ray/ml/train/integrations/torch/torch_trainer.py

```diff
@@ -0,0 +1,54 @@
+from typing import Callable, Optional, Dict
+
+from ray.train.torch import TorchConfig
+from ray.ml.trainer import GenDataset
+from ray.ml.train.impl.data_parallel_trainer import DataParallelFunctionTrainer
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.util import PublicAPI
+
+
+@PublicAPI(stability="alpha")
+class TorchTrainer(DataParallelFunctionTrainer):
+    """A Trainer for data parallel PyTorch training.
+
+    This Trainer runs the provided function on multiple Ray Actors. The
+    provided datasets are automatically sharded to each Ray actor running
+    the training function.
+
+    Args:
+        train_func (Callable): The training function to execute.
+            This can either take in no arguments or a ``config`` dict.
+        train_func_config (Optional[Dict]): Configurations to pass into
+            ``train_func`` if it accepts an argument.
+        torch_config (Optional[TorchConfig]): Configuration
+            for setting up the PyTorch backend. If set to None, use the
+            default configuration.
+        scaling_config: Configuration for how to scale training.
+        run_config: Configuration for the execution of the training run.
+        train_dataset: Either a distributed Ray :ref:`Dataset <dataset-api>`
+            or a Callable that returns a Dataset, to use for training. If a
+            ``preprocessor`` is also provided, it will be fit on this
+            dataset and this dataset will be transformed.
+        extra_datasets: Any extra Datasets (such as validation or test
+            datasets) to use for training. If a ``preprocessor`` is
+            provided, the datasets specified here will only be transformed,
+            and not fit on.
+        preprocessor: A preprocessor to preprocess the provided datasets.
+        resume_from_checkpoint: A checkpoint to resume training from.
+    """
+
+    def __init__(
+        self,
+        train_func: Callable,
+        train_func_config: Optional[Dict] = None,
```

Comment:
If we pass these two into the constructor now, instead of using the original `train_function` method, then why not do it like this in the first place (i.e. on the interface)?

I think the downside of this is that users now have to be very careful with the signature of their function.

Another drawback is that this function does not have access to other attributes this way, like `train_dataset`.
I'm not convinced this works (would be great if it did, though!)

Reply:
This is a good point- we decided to go with this for now, but can revisit in the future.